### ISSUE TITLE
add a function to check for checksummed addresses

### ISSUE THREAD

**Issue by ProofOfKeags** [f-o-a-m/hs-web3#11]
So all of the internals interpret the address as bytes. Ideally before we parse the address from the string, we'd have some function to see if the address checksums. I know before you wanted to try and implement a parse function that was like:

```haskell
parseAddress :: Text -> Either Text Web3.Address
```
If this existed, perhaps we could add a different one that was

```haskell
parseAddressChecksummed :: Text -> Either Text Web3.Address
```

or make it a boolean parameter and unify them.

Thoughts?

**Comment by martyall** [comment]
I think having a separate checksummed parser is a good idea rather than boolean flag. 

That parse function is now from `Bytestring` and we just re-exposed it. This is definitely more convenient to use than `Bytes 160`. If we want to make versions from `Text`, that's fine too. I don't have a problem keeping both.

In almost all application code that we write at FOAM, the internal representation for Address is basically text

**Comment by cmditch** [comment]
FYI, here's some FP algos for dealing with ethereum address checksumming. Hope it's useful.

https://github.com/cmditch/elm-ethereum/blob/master/src/Eth/Utils.elm#L415
https://github.com/cmditch/elm-ethereum/blob/master/src/Eth/Utils.elm#L180

**Comment by ProofOfKeags** [comment]
Totally cool with ByteString instead of Text. I'll take a look at @cmditch 's algos and put in a PR.

**Comment by ProofOfKeags** [comment]
Is the ByteString you're parsing from already in Base16 or is it parsing actual utf8 encoded hex characters

**Comment by martyall** [comment]
https://github.com/f-o-a-m/hs-web3/blob/master/src/Network/Ethereum/ABI/Prim/Address.hs#L56

The input to this function is assumed to be Base16 encoded.

Likewise if you turn it back into a bytestring via this function

https://github.com/f-o-a-m/hs-web3/blob/master/src/Network/Ethereum/ABI/Prim/Address.hs#L63

then the resulting bytestring is Base16 encoded

**Comment by ProofOfKeags** [comment]
I think I have a sensible implementation for this. The unfortunate thing though is that since the Address type is stored in byte form, I had to operate over raw ByteString's for this thing. I could have newtyped it or possibly used the BytesN 42 to ensure the length was right. All that said. The functions I defined have undefined behavior for ByteStrings that aren't valid Base16 encoded Addresses with the "0x" prefix. I feel like this is reasonable, but I'd like to get some feedback.
